hydra/controllers/client.py

```diff
@@ -816,3 +817,67 @@ def disable_metrics(self):
 
         self.restart_service()
         self.app.log.info(f'Successfully disabled all metrics reporting.')
+
+    @ex(
+        arguments=[
+            (
+                ['-n', '--name'],
+                {
+                    'help': 'name of network to join',
+                    'action': 'store',
+                    'dest': 'name'
+                }
+            ),
+            (
+                ['-v', '--version'],
+                {
+                    'help': 'version of binary',
+                    'action': 'store',
+                    'dest': 'version'
+                }
+            ),
+            (
+                ['-D', '--destroy'],
+                {
+                    'help': 'destroy existing directory',
+                    'action': 'store_true',
+                    'dest': 'destroy'
+                }
+            ),
+        ]
+    )
+    def upgrade_binary(self, version=None, destroy=False):
+        name = self.app.utils.env_or_arg(
+            'name', 'HYDRA_NETWORK', or_path='.hydra_network', required=True)
+
+        destination = self.app.utils.path(name)
+
+        service_name = f'{name}.service'
+        systemd_service = f'/etc/systemd/system/{service_name}'
+
+        os.chdir(destination)
+
+        if self.app.pargs.version:
+            version = self.app.pargs.version
+
+        if os.path.exists('shipchain'):
+            if not self.app.pargs.destroy:
+                self.app.log.info(f'Binary file exists, destroy and replace with -D')
+                return
+            os.remove('shipchain')
```

Comment:
The deletion of the binary should happen after the download is successful. Also I'm not sure if this would work here since the service is still running.

It should go

1. Download new binary to temp file (`shipchain-temp` or `/tmp/shipchain`)
2. If download successful, stop service
3. Overwrite old binary with new binary
4. Start service

Reply:
Thanks, that makes sense. Done.